File: skills/tokensmind-agent-network-runtime/scripts/lib/runtime/crypto.js
import { createHash, randomBytes, randomUUID } from 'node:crypto';

function randomCredential(prefix) {
  return `${prefix}${randomBytes(32).toString('base64url')}`;
}

export function sha256(value) {
  return createHash('sha256').update(value, 'utf8').digest('hex');
}

export function createAuthorizationMaterial({ devicePrefix, tokenPrefix }) {
  const deviceCode = randomCredential(devicePrefix);
  const token = randomCredential(tokenPrefix);

  return {
    authorizationId: randomUUID(),
    createIdempotencyKey: randomUUID(),
    exchangeIdempotencyKey: randomUUID(),
    deviceCode,
    deviceCodeHash: sha256(deviceCode),
    token,
    tokenHash: sha256(token),
    tokenPrefix: token.slice(0, 20),
  };
}

File: skills/tokensmind-agent-network/scripts/python_runtime/crypto.py
import base64
import hashlib
import secrets
import uuid


def sha256(value):
    return hashlib.sha256(value.encode("utf-8")).hexdigest()


def _random_credential(prefix):
    encoded = base64.urlsafe_b64encode(secrets.token_bytes(32)).decode("ascii").rstrip("=")
    return prefix + encoded


def create_authorization_material(device_prefix, token_prefix):
    device_code = _random_credential(device_prefix)
    token = _random_credential(token_prefix)
    return {
        "authorizationId": str(uuid.uuid4()),
        "createIdempotencyKey": str(uuid.uuid4()),
        "exchangeIdempotencyKey": str(uuid.uuid4()),
        "deviceCode": device_code,
        "deviceCodeHash": sha256(device_code),
        "token": token,
        "tokenHash": sha256(token),
        "tokenPrefix": token[:20],
    }
